explorer/chartscache.go

```diff
@@ -124,3 +126,278 @@ func ReadCacheFile(filePath string, height int64) error {
+		cacheChan: reorgChan,
+	}
+}
+
+// ReorgHandler handles the charts cache data reorganization.
+func (m *ChainMonitor) ReorgHandler() {
+	defer m.wg.Done()
+	for {
+		select {
+		case data, ok := <-m.cacheChan:
+			if !ok {
+				return
+			}
+
+			// Drop all entries since since the ancestor block after reorganization
+			// of all PG and Sqlite DB is complete.
+			if err := m.explorer.DropToAncestor(uint64(data.NewChainHeight)); err != nil {
+				log.Errorf("cache reorg failed: %v", err)
+				return
+			}
+
+			// Initiate cache update after drop the entries till the ancestor bloxk.
+			m.explorer.prePopulateChartsData()
+
+			data.WG.Done()
+
+		case <-m.ctx.Done():
+			return
+		}
+	}
+}
+
+// DropToAncestor drops all the chart data entries since the common ancestor
+// block. If the ancestor block is in the last entry added, then the last chart
+// data entry is the only one that is dropped.
+func (exp *explorerUI) DropToAncestor(heightAfterAncestor uint64) error {
+	timeAfterAncestor, err := exp.explorerSource.BlockTimeByHeight(int64(heightAfterAncestor))
+	if err != nil {
+		return err
+	}
+
+	cacheData := cacheChartsData.get()
+
+	// "avg-block-size" chart data.
+	data, ok := cacheData[dbtypes.AvgBlockSize]
+	if ok {
+		var index = len(data.Time) - 1
+		for ; index > 0; index-- {
+			if data.Time[index].UNIX() < timeAfterAncestor {
+				break
+			}
+		}
+		cacheData[dbtypes.AvgBlockSize].Time = cacheData[dbtypes.AvgBlockSize].Time[:index]
+		cacheData[dbtypes.AvgBlockSize].Size = cacheData[dbtypes.AvgBlockSize].Size[:index]
+	}
+
+	// "blockchain-size" chart data.
+	data, ok = cacheData[dbtypes.BlockChainSize]
+	if ok {
+		var index = len(data.Time) - 1
+		for ; index > 0; index-- {
+			if data.Time[index].UNIX() < timeAfterAncestor {
+				break
+			}
+		}
+		cacheData[dbtypes.BlockChainSize].Time = cacheData[dbtypes.BlockChainSize].Time[:index]
+		cacheData[dbtypes.BlockChainSize].ChainSize = cacheData[dbtypes.BlockChainSize].ChainSize[:index]
+	}
+
+	// "chainwork" chart data.
+	data, ok = cacheData[dbtypes.ChainWork]
+	if ok {
+		var index = len(data.Time) - 1
+		for ; index > 0; index-- {
+			if data.Time[index].UNIX() < timeAfterAncestor {
+				break
+			}
+		}
+		cacheData[dbtypes.ChainWork].Time = cacheData[dbtypes.ChainWork].Time[:index]
+		cacheData[dbtypes.ChainWork].ChainWork = cacheData[dbtypes.ChainWork].ChainWork[:index]
+	}
+
+	// "coin-supply" chart data.
+	data, ok = cacheData[dbtypes.CoinSupply]
+	if ok {
+		var index = len(data.Time) - 1
+		for ; index > 0; index-- {
+			if data.Time[index].UNIX() < timeAfterAncestor {
+				break
+			}
+		}
+		cacheData[dbtypes.CoinSupply].Time = cacheData[dbtypes.CoinSupply].Time[:index]
+		cacheData[dbtypes.CoinSupply].ValueF = cacheData[dbtypes.CoinSupply].ValueF[:index]
+	}
+
+	// "duration-btw-blocks" chart data.
+	data, ok = cacheData[dbtypes.DurationBTW]
+	if ok {
+		var index = len(data.Height) - 1
+		for ; index > 0; index-- {
+			if data.Height[index] < heightAfterAncestor {
```

Comment:
Based on these conditions, the loop will break when `data.Height[index]` is equal to ancestor height.  I would think it would be easier to reason about the code if it just used `<=` with the appropriate height condition.  When the loop breaks, the index of the ancestor is used as the end slice index (`Height[:index]`), so it indeed is removing the ancestor.  That is OK, although theoretically unnecessary if it is the _common_ ancestor, but it does contradict the wording "drops all the chart data entries since the common ancestor" in the function docs.